src/main/java/io/confluent/connect/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -365,6 +383,31 @@ private BulkResponse execute() throws Exception {
 
   }
 
+  /**
+   * Compute the time to sleep using exponential backoff with jitter. This method computes the normal exponential backoff
+   * based upon the {@code minRetryBackoffMs} and the {@code retryAttempt}, bounding the result to be within {@code minRetryBackoffMs}
+   * and {@code maxRetryBackoffMs}, and then choosing a random value within that range.
+   * <p>
+   * The purposes of using exponential backoff is to give the ES service time to recover when it becomes overwhelmed.
+   * Adding jitter attempts to prevent a thundering herd, where large numbers of requests from many tasks overwhelm the
+   * ES service, and without randomization all tasks retry at the same time. Randomization should spread the retries
+   * out and should reduce the overall time required to complete all attempts.
+   * See <a href="https://www.awsarchitectureblog.com/2015/03/backoff.html">this blog post</a> for details.
+   *
+   * @param retryAttempt the retry attempt number, starting with 1 for the first retry
+   * @param minRetryBackoffMs the minimum time to wait before retrying; assumed to be 0 if value is negative
+   * @param maxRetryBackoffMs the maximum amount of time to wait before retrying
+   * @return the time in milliseconds to wait before the next retry attempt, in the range {@code minRetryBackoffMs}
+   * and {@code maxRetryBackoffMs} (inclusive), or {@code minRetryBackoffMs} if {@code minRangeBackoffMs}
+   * is greater than or equal to {@code maxRetryBackoffMs}
+   */
+  protected static long computeRetryWaitTimeInMillis(int retryAttempt, long minRetryBackoffMs, long maxRetryBackoffMs) {
+    if (minRetryBackoffMs >= maxRetryBackoffMs) return minRetryBackoffMs; // this was the original value and likely what they're setting
+    if (minRetryBackoffMs < 0) minRetryBackoffMs = 0;
+    long nextMaxTimeMs = Math.max(minRetryBackoffMs + 1, Math.min(maxRetryBackoffMs, minRetryBackoffMs * 2 ^ retryAttempt));
```

Comment:
In java `^` corresponds to XOR and not `Math.pow`. This makes me think we probably need a test to confirm that the actual waiting time corresponds to exponential backoff. 

Since we are raising at a power of 2, I'd suggest bit shifting, such as: `minRetryBackoffMs << retryAttempt;` (haven't tried it, please confirm, we'll be doing something similar in the S3 connector).

Reply:
Yeah, not sure what I was thinking. Thank goodness for reviews!